Competition2018/src/main/java/competition/subsystems/drive/commands/AssistedTankDriveCommand.java

```diff
@@ -0,0 +1,51 @@
+package competition.subsystems.drive.commands;
+
+import com.google.inject.Inject;
+
+import competition.operator_interface.OperatorInterface;
+import competition.subsystems.drive.DriveSubsystem;
+import xbot.common.command.BaseCommand;
+import xbot.common.injection.wpi_factories.CommonLibFactory;
+import xbot.common.math.PIDFactory;
+import xbot.common.math.PIDManager;
+import xbot.common.math.XYPair;
+import xbot.common.subsystems.drive.control_logic.HeadingAssistModule;
+import xbot.common.subsystems.drive.control_logic.HeadingModule;
+
+public class AssistedTankDriveCommand extends BaseCommand {
+
+    final DriveSubsystem drive;
+    final OperatorInterface oi;
+    final HeadingAssistModule ham;
+    
+    @Inject
+    public AssistedTankDriveCommand(CommonLibFactory clf, PIDFactory pf, DriveSubsystem drive, OperatorInterface oi) {
+        this.drive = drive;
+        this.oi = oi;
+        this.requires(drive);
+        
+        PIDManager headingPid = pf.createPIDManager("DriveHeading", 4, 0, 0);
+        PIDManager decayPid = pf.createPIDManager("DriveDecay", 0, 0, 1);
+        
+        HeadingModule holdHeading = clf.createHeadingModule(headingPid);
+        HeadingModule decayHeading = clf.createHeadingModule(decayPid);
+        ham = clf.createHeadingAssistModule(holdHeading, decayHeading);
+    }
+    
+    @Override
+    public void initialize() {
+        log.info("Initializing");
+    }
+
+    @Override
+    public void execute() {
+        double left = oi.gamepad.getLeftVector().y; 
+        double right = oi.gamepad.getRightVector().y;
+        
+        double yTranslate = (left+right) / 2;
+        double turn = ham.calculateHeadingPower((right-left)/2);
+        
+        drive.drive(new XYPair(0, yTranslate), turn);
```

Comment:
Could you look into writing a simple proof that this is equivalent to standard tank drive? I spent some time thinking about it and it seems that the heading conversion with be necessarily different with the current implementation because a rational factor is used for a "heading power" rather than a rotational Omega value.

Reply:
Can you give a quick example of a leftPower and rightPower that won't convert well?

Reply:
``` 
L = (L+R)/2 - (R-L)/2
2L = L+R-R+L
2L = 2L
L = L

R = (L+R)/2 + (R-L)/2
2R = L+R+R-L
2R = 2R
R = R
```

Reply:
Ah wait... after re-reading a bit, I think you're asking something different about the interaction between the heading assist module and the tank drive? I'm not certain.

Reply:
The left wheel's power in a normal tank drive system would be calculated as:
```
V_l=K_l*J_l
```
where V_I is output power, J_I is input from the left joystick and K_l is any scaling factor you may use (typically 1). That's fairly obvious. Now, if we do the equivalent derivation for the new drive subsystem using the math as it currently stands, I get (feel free to check this):

![ql_838d6e5b69e54e355647785c40f6c1d7_l3](https://user-images.githubusercontent.com/3310349/35358088-582580ee-010a-11e8-83d1-3bd55c0a61c3.png)

... where N is the normalization factor.

Notice that, aside from the case where `K_w = K_y` for both wheels, each depends on the input of both joysticks. This only collapses to be equivalent to the earlier case when the aforementioned constants are equal and N = 1. Hence, we know that this system is _not_ equivalent to typical tank drive aside from the special case. I haven't looked into the particulars of its properties in relation to the constants.

Reply:
Based on this, it seems to me that the new drive subsystem implementation _can_ work for tank drive, but it should have a superclass or other drive type-specific interface such that it prevents you from using non-equal y and w constants to prevent this unexpected behavior.

Reply:
Re: your last sentence - are you saying that if the various `drive()` methods constrained their inputs to be in robot scale (-1 to 1) we'd be good?

Reply:
Sorry, I meant `K_y` and `K_w` from my earlier post -- the `y` and `w` constants in a `MotorRegistration`. For tank drive to operate as expected, all wheels must share the same K_iy value and all must share the same K_iw value, _and_ that K_iw value must be equal to K_iy. Also, to avoid the normalization factor causing dead zones at the extremes of joystick input, K_iy +K_iw = 0.5.

The only combination of factors that satisfies those requirements would be K_y=K_w=0.25